Add [Dwarf_low] library

mshinwell opened the pull request:
This pull request adds a low-level library for the writing of DWARF information.  Currently constructs in DWARF version 4 plus some GNU extensions are supported.  A higher-level library will follow in due course which provides various abstractions that make it very much easier to emit correct DWARF.

@bschommer Perhaps you could eyeball this, based on your previous offer?  I'd like it to land in 4.08 if at all possible so we are a bit short on time.  Please see the note on code review scrutiny in #2069.  Please also note that this branch is based on previous pull requests, so just look at the individual relevant commits, rather than the whole thing.  Each commit relevant to this pull request will be marked with its number (there is currently only one such commit).

mshinwell committed:
GPR#2073 (Asm_directives and Target_system)

mshinwell committed:
GPR#2077 (Dwarf_low library)

mshinwell committed:
GPR#2073: remove private from a type

bschommer commented:
I will have  a look at it.

mshinwell committed:
GPR#2077: I hate integers

mshinwell commented:
I realised that this library still needed some work to correctly distinguish the various types of integers involved.  I hate integers.  As if there weren't enough with host and target integers in all their different varieties, we now have "DWARF width" integers as well, which are either 32-bit or 64-bit depending on which DWARF format we are emitting (which will be controlled by a command-line switch).  This should be dealt with by `204c561`, which hasn't been tested yet, but will be once I finish the library that sits above this over the next couple of days.

mshinwell commented:
(`204c561` also sneaks in a change to consistently use `Linearize.label` rather than `Cmm.label`.)

bschommer commented:
You can actually avoid a lot of the computations if you use the assembler for size calculation like for example gcc does, by emitting labels and using the difference between the labels as size.

mshinwell commented:
I do use the distance-between-labels technique in some places, but I think it works quite well to define the size calculation manually for the basic structures.  As far as I recall there are some cases where the assembler would not be able to encode the distances in any case (since they require variable-length encoding).

mshinwell committed:
GPR#2077: Integer size fix

mshinwell commented:
Not that I want to change the fundamental method of operation of the code in any case, but I've realised that I misspoke in the above comment: I was forgetting that we don't manually encode the variable-length integers and in fact this is supported by the assembler.  (The variable-length calculation we have is just for the sizing.)

bschommer commented on asmcomp/debug/dwarf_low/dwarf_version.ml:
```diff
@@ -0,0 +1,44 @@
+[@@@ocaml.warning "+a-4-30-40-41-42"]
+
+type t =
+  | Dwarf_2
```
Since you emit Dwarf 4 anyway, why adding Dwarf 2 and Dwarf 3?

mshinwell replied:
Just reflecting the specification (see comment below).

bschommer commented on asmcomp/debug/dwarf_low/debug_info_section.ml:
```diff
@@ -0,0 +1,68 @@
+  Dwarf_version.emit dwarf_version;
+  Dwarf_value.emit (debug_abbrev_offset t);
+  Dwarf_value.emit address_width_in_bytes_on_target;
+  List.iter (fun die -> DIE.emit die) t.dies
```
I think there is a final 0 missing, to indicate the end of the debug_info section

mshinwell replied:
I can't find any reference in the DWARF-4 spec to such a null byte.  Do you have a reference?

bschommer replied:
You are right, there doesn't need to be a null byte.

bschommer commented:
I started my review, overall it looks good, but I need some time to check some of the values. One suggestion which would be to add comments in the emitted code for example with name of attributes, abbreviation values, etc. as for example clang or CompCert does, which makes later debugging easier.

bschommer commented on asmcomp/debug/dwarf_low/dwarf_operator.ml:
```diff
@@ -0,0 +1,595 @@
+type t =
+  | DW_op_addr of string
+  | DW_op_regx of { reg_number : int; }
+  | DW_op_reg0
```
You could simplify the code, by not differentiating between `DW_op_regx` and `DW_op_reg0` to `DW_op_reg31` and just adding a simple check for in `emit`, `print` and `encode` if the register number is between 0 and 31.

mshinwell replied:
Ah, so there is a design point here.  The types are supposed to reflect the specification very closely, which is why everything is there.  The next level up (`Dwarf_high`) deals with transformations such as the ones you suggest.

Apart from being a general point of principle, this is also a potentially useful property if this library is in the future enhanced to read DWARF, which it may well be.

bschommer replied:
Was just a suggestion for easing the implementation, but this is also actually fine.

bschommer commented on asmcomp/debug/dwarf_low/dwarf_operator.ml:
```diff
@@ -0,0 +1,595 @@
+  | DW_op_reg30
+  | DW_op_reg31
+  | DW_op_fbreg of { offset_in_bytes : Targetint.t; }
+  | DW_op_breg0 of { offset_in_bytes : Targetint.t; }
```
This could be simplified similar to the `DW_op_reg`, see comment above.

mshinwell replied:
See above.

mshinwell commented:
I will see about adding comments in the assembly output.  I think I hacked that up for debugging once, but it isn't in this version.

bschommer commented on asmcomp/debug/dwarf_low/dwarf_operator.ml:
```diff
@@ -0,0 +1,595 @@
+  | DW_op_breg29 of { offset_in_bytes : Targetint.t; }
+  | DW_op_breg30 of { offset_in_bytes : Targetint.t; }
+  | DW_op_breg31 of { offset_in_bytes : Targetint.t; }
+  | DW_op_bregx of { reg_number : int; offset_in_bytes : Targetint.t; }
```
The reg number should be an ULEB128

mshinwell replied:
I don't have a type specifically for numbers that are variable-length encoded.  I think `int` is very likely large enough here in all cases.  I have added a comment.

bschommer commented on asmcomp/debug/dwarf_low/dwarf_operator.ml:
```diff
@@ -0,0 +1,595 @@
+  | DW_op_bregx of { reg_number : int; offset_in_bytes : Targetint.t; }
+  | DW_op_deref
+  | DW_op_plus_uconst of Targetint.t
+  | DW_op_consts of Targetint.t
```
The different size variants of the consts are missing, even if they are not needed, a comment why they are excluded would be nice. Also some of the other operands which are probably not needed are missing.

mshinwell replied:
Fixed, see comment below that I am about to write.

bschommer commented on asmcomp/debug/dwarf_low/aranges_table.ml:
```diff
@@ -0,0 +1,55 @@
+    V.Code_address_from_symbol_diff
+      { upper = end_of_code_symbol; lower = start_of_code_symbol; };
+    V.Absolute_code_address Targetint.zero;
+    V.Absolute_code_address Targetint.zero;
```
I think there is a zero missing, the standard says:

     A particular set is terminated by an entry consisting of three zeroes.

mshinwell replied:
Ah, except that since we selected "flat address space" just above, then I think the terminating entry is only two zeroes (see the next sentence in the standard).  I have added some more comments in the code.

bschommer replied:
Indeed.

bschommer commented on asmcomp/debug/dwarf_low/dwarf_attributes.ml:
```diff
@@ -0,0 +1,398 @@
+    | Ref2 : (Class.reference, [< ref1 | ref2 ]) t
+    | Ref4 : (Class.reference, [< ref1 | ref2 | ref4 ]) t
+    | Ref8 : (Class.reference, [< ref1 | ref2 | ref4 | ref8 ]) t
+    | Ref_udata : (Class.reference, ref_udata) t
```
The `DW_FORM_indirect` is missing, I don't think you need it but maybe a comment why it is not included would be helpful.

mshinwell replied:
Done.

bschommer commented on asmcomp/debug/dwarf_low/dwarf_operator.ml:
```diff
@@ -0,0 +1,595 @@
+
+type t =
+  | DW_op_addr of string
+  | DW_op_regx of { reg_number : int; }
```
The reg_number should be a ULEB128.

mshinwell replied:
Same as above, comment added.

bschommer commented:
I finished my first look at all files in dwarf_low. Aside from the few comments I could not find any obvious mistakes.

mshinwell committed:
GPR#2077: Code review; rewrite [Dwarf_operator]

mshinwell commented:
@bschommer I hadn't realised so many operators were missing.  I've now added all of the DWARF-4 ones, I think.  At the same time I realised that the printing, size calculation and emitting code could be functorised, which substantially improves maintainability I think.  I haven't yet re-tested this but will do so in the coming days.  Comments on this appreciated.  Sorry that many of the cases have been re-ordered; I did that to bring everything into line with the main textual description of the operators in the standard (rather than the ordering in which they are described in the data representation part of the standard).

mshinwell commented:
As a note, the intention is that this library (and `Dwarf_high`) will use the new `Asm_label` and `Asm_symbol` modules from #2073 throughout.  I am preparing a patch for that, but it won't be needed until #2073 has been merged.

mshinwell commented:
I've added support for properly commented DWARF output in the `.s` files (when `-S` is specified to the compiler) and also introduced some new unsigned integer types, since we were falling foul of the bounds checks in e.g. `Numbers.Int8` at some points.  I'm not going to put these changes onto this PR until #2073 has been merged, since disentangling them will be tricky.

In the meantime, here's what the output looks like.  Header for the `.debug_info` section:
```assembly
        .section .debug_info
.L127:
        .long   0x3e85  /* 32-bit initial length */
        .short  0x4     /* DWARF version */
        .long   .L11025 /* abbrevs. for this comp. unit (offset into .debug_abbrev) */
        .byte   0x8     /* Arch.size_addr */
```
Debugging information entries in the section (in this case for the overall compilation unit and the module initialiser):
```assembly
.L125:
        .uleb128 0x1    /* abbrev. code (abbrev. has tag DW_TAG_compile_unit) */
        .long   .L11028 /* name (DW_AT_name : DW_FORM_strp) (offset into .debug_str) */
                        /*   (.debug_str entry starts "/usr/local/home/mshinwell/mshi" [...]) */
        .long   .L124   /* (DW_AT_stmt_list : DW_FORM_sec_offset_lineptr) (offset into .debug_line) */
        .quad   camlList__code_begin    /* low PC value (DW_AT_low_pc : DW_FORM_addr) */
        .quad   camlList__code_end      /* high PC value (DW_AT_high_pc : DW_FORM_addr) */
        .long   .L11031 /* compilation directory (DW_AT_comp_dir : DW_FORM_strp) (offset into .debug_str) */
                        /*   (.debug_str entry starts "/usr/local/home/mshinwell/mshi" [...]) */
        .long   .L11033 /* producer name (DW_AT_producer : DW_FORM_strp) (offset into .debug_str) */
                        /*   (.debug_str entry starts "ocamlopt 4.08.0+dev0-2018-04-0" [...]) */
.L10860:
        .uleb128 0x2    /* abbrev. code (abbrev. has tag DW_TAG_subprogram) */
        .long   .L10859 /* ref. to DIE (DW_AT_sibling : DW_FORM_ref_addr) (offset into .debug_info) */
        .long   .L11036 /* name (DW_AT_name : DW_FORM_strp) (offset into .debug_str) */
                        /*   (.debug_str entry is "List") */
        .quad   camlList__entry /* low PC value (DW_AT_low_pc : DW_FORM_addr) */
        .quad   .L10849 /* high PC value (DW_AT_high_pc : DW_FORM_addr) */
                        /* visible externally (DW_AT_external : DW_FORM_flag_present) (Flag_true elided) */
        .long   .L10859 /* ref. to DIE (DW_AT_type : DW_FORM_ref_addr) (offset into .debug_info) */
.L10861:
```
The DWARF attributes and forms are printed alongside the attribute values.  This means you don't need to look up the abbreviation code in the `.debug_abbrev` section to find out which attributes values are being given for.  (Awesome.)

Note how strings which live in the `.debug_str` section are printed in the comments, so you don't need to go looking for them.

The abbreviations table itself looks like this:
```assembly
        .uleb128 0x1    /* abbrev. code (abbrev. has tag DW_TAG_compile_unit) */
        .uleb128 0x11   /* DW_TAG_compile_unit */
        .byte   0x1     /* Child determination */
        .uleb128 0x3    /* DW_AT_name */
        .uleb128 0xe    /* DW_FORM_strp */
        .uleb128 0x10   /* DW_AT_stmt_list */
        .uleb128 0x17   /* DW_FORM_sec_offset_lineptr */
        .uleb128 0x11   /* DW_AT_low_pc */
        .uleb128 0x1    /* DW_FORM_addr */
        .uleb128 0x12   /* DW_AT_high_pc */
        .uleb128 0x1    /* DW_FORM_addr */
        .uleb128 0x1b   /* DW_AT_comp_dir */
        .uleb128 0xe    /* DW_FORM_strp */
        .uleb128 0x25   /* DW_AT_producer */
        .uleb128 0xe    /* DW_FORM_strp */
        .uleb128 0x0    /* terminator word 1 */
        .uleb128 0x0    /* terminator word 2 */

        .uleb128 0x2    /* abbrev. code (abbrev. has tag DW_TAG_subprogram) */
        .uleb128 0x2e   /* DW_TAG_subprogram */
        .byte   0x1     /* Child determination */
        .uleb128 0x1    /* DW_AT_sibling */
        .uleb128 0x10   /* DW_FORM_ref_addr */
        .uleb128 0x3    /* DW_AT_name */
        .uleb128 0xe    /* DW_FORM_strp */
        .uleb128 0x11   /* DW_AT_low_pc */
        .uleb128 0x1    /* DW_FORM_addr */
        .uleb128 0x12   /* DW_AT_high_pc */
        .uleb128 0x1    /* DW_FORM_addr */
        .uleb128 0x3f   /* DW_AT_external */
        .uleb128 0x19   /* DW_FORM_flag_present */
        .uleb128 0x49   /* DW_AT_type */
        .uleb128 0x10   /* DW_FORM_ref_addr */
        .uleb128 0x0    /* terminator word 1 */
        .uleb128 0x0    /* terminator word 2 */
```

Part of a location list, showing a variable which started out life in a register, and was then spilled.  The expression written in the DWARF stack machine language is shown in readable form:
```assembly
.L2392:
                        /* Base address selection entry: */
        .quad   0xffffffffffffffff      /* largest representable addr. offset */
        .quad   camlList__rmap2_f_629   /* base address */

                        /* Location list entry: */
        .quad   .L2278 - camlList__rmap2_f_629  /* beginning address */
        .quad   .L2279 - camlList__rmap2_f_629  /* ending address */
        .short  0x1     /* expression size */
                        /* Single location description (DWARF expression): */
        .byte   0x50    /* DW_op_reg0 */

                        /* Location list entry: */
        .quad   .L2279 - camlList__rmap2_f_629  /* beginning address */
        .quad   .L2296 - camlList__rmap2_f_629  /* ending address */
        .short  0x4     /* expression size */
                        /* Single location description (DWARF expression): */
        .byte   0x9c    /* DW_op_call_frame_cfa */
        .byte   0x11    /* DW_op_consts */
        .sleb128 0x20   /*   arg of DW_OP_consts */
        .byte   0x1c    /* DW_op_minus */
        .quad   0x0     /* end marker */
        .quad   0x0     /* end marker */
```

bschommer commented:
Overall the comments look good and are verbose enough for debugging. I will have a look at the rest of the changes.

bschommer commented:
I had a look at the new constructs introduced in the last commit and it looks good to me.

mshinwell committed:
Move dwarf_low/ into asmcomp/debug/dwarf/

mshinwell committed:
Import latest dwarf_low/ changes from gdb-names-gpr branch

mshinwell commented:
@bschommer I've pushed a changeset (`e4d08f5`) that contains my latest changes to this library.  These are sufficient to get quite respectable debugging support in gdb, which has been tested both on Linux and macOS (yes, it actually works on macOS).  The previous changeset just consists of renames (this is because I am splitting the DWARF-specific debug information generation code from the generic debug passes).  I would appreciate it if you could find the time to cast an eye over these changes.

This code contains support for DWARF 5.  The DWARF 4-specific modules have been moved into a subdirectory.  I'm afraid there isn't any git tracking of those moves, so the best thing to do is probably to use the Jane Street `patdiff` tool or equivalent to look at these before and after `e4d08f5`.  I thought DWARF 5 support was going to be required to get things such as call site marking to work, although it turns out that this can be done with DWARF 4 GNU extensions.  Nonetheless, the support is there now, and will become important in the future as toolchains start to fully support DWARF 5.  The OCaml compiler's default will be DWARF 4 + GNU extensions for now.

bschommer commented:
Thanks for the update, I will have a look at this in the next days.

Since I can no longer add new Comments, I will just edit this:
I tried to build this branch locally and it failed to build, since the file asmcomp/debug/dwarf_low/dwarf_emittable.cmo could not be found, I think there might be some missing change in the Makefile.

bschommer commented:
I had a first look at the changes and overall could not find any obvious flaw. I'm not sure if it is really effective to generate the comments and just not emit then, since they still cause quite some allocations. Concerning correct generated debug information, have you considered using ````dwarfdump```` for testing purposes, it actually is quite helpful and also suggests some possible size optimizations.

mshinwell commented:
@bschommer This PR to be returned to, hopefully later this year -- thanks for the review.
